versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -0,0 +1,936 @@
+                .filter(kt -> seen.add(kt.getKey()))
+                .filter(kt -> keyFilter.check(kt.getKey(), kt.getContentsId(), kt.getType()))
+                .forEach(keys::add);
+          }
+          e.getPuts().stream()
+              .filter(kt -> seen.add(kt.getKey()))
+              .filter(kt -> keyFilter.check(kt.getKey(), kt.getContentsId(), kt.getType()))
+              .map(KeyWithBytes::asKeyWithType)
+              .forEach(keys::add);
+          // Only have 'Key' in deletes, so map it to a 'KeyWithType' and it can be used to remove
+          // the entry from 'keys'.
+          e.getDeletes().stream().filter(seen::add).forEach(removes::add);
+        });
+
+    return keys.stream().filter(kt -> !removes.contains(kt.getKey()));
+  }
+
+  /**
+   * Fetch the global-state and per-ref contents for the given {@link Key}s and {@link Hash
+   * commitSha}. Non-existing keys must not be present in the returned map.
+   */
+  protected Map<Key, ContentsAndState<ByteString>> fetchValues(
+      OP_CONTEXT ctx, Hash refHead, List<Key> keys, KeyFilterPredicate keyFilter) {
+    Set<Key> remainingKeys = new HashSet<>(keys);
+
+    Map<Key, ByteString> nonGlobal = new HashMap<>();
+    Map<Key, String> keyToContentsIds = new HashMap<>();
+    Set<String> contentsIds = new HashSet<>();
+    try (Stream<CommitLogEntry> log =
+        takeUntilExcludeLast(commitLogFetcher(ctx, refHead), e -> remainingKeys.isEmpty())) {
+      log.peek(entry -> entry.getDeletes().forEach(remainingKeys::remove))
+          .flatMap(entry -> entry.getPuts().stream())
+          .filter(put -> remainingKeys.remove(put.getKey()))
+          .filter(put -> keyFilter.check(put.getKey(), put.getContentsId(), put.getType()))
+          .forEach(
+              put -> {
+                nonGlobal.put(put.getKey(), put.getValue());
+                keyToContentsIds.put(put.getKey(), put.getContentsId());
+                contentsIds.add(put.getContentsId());
+              });
+    }
+
+    Map<String, ByteString> globals = fetchGlobalStates(ctx, contentsIds);
+
+    return nonGlobal.entrySet().stream()
+        .collect(
+            Collectors.toMap(
+                Entry::getKey,
+                e ->
+                    ContentsAndState.of(
+                        e.getValue(), globals.get(keyToContentsIds.get(e.getKey())))));
+  }
+
+  /**
+   * Fetches the global-state information for the given content-ids.
+   *
+   * @param ctx technical context
+   * @return map of content-id to state
+   */
+  protected abstract Map<String, ByteString> fetchGlobalStates(
+      OP_CONTEXT ctx, Set<String> contentIds);
+
+  /**
+   * Write a new commit-entry with a best-effort approach to prevent hash-collisions but without any
+   * other consistency checks/guarantees. Some implementations however can enforce strict
+   * consistency checks/guarantees.
+   */
+  protected abstract void writeIndividualCommit(OP_CONTEXT ctx, CommitLogEntry entry)
```

Comment:
Yes, implementations have to rather "blindly" write the given entry. No value or consistency checks by the implementations. Implementations however _can optionally_ ensure that the given hash is not already present - but since it might be tricky or quite expensive for some databases, it's not mandatory.

Clarified that in the javadoc.

Reply:
Ok, maybe I am confused. Is this the update of the ref or just the write of the commit w/ hash? The commit write cna be blind but the ref swap can't. Did I just misunderstand where the ref swap happens?

Reply:
Yea - this method just blindly writes the commit. The actual swap-operation depends on the database, so it's not in this PR.